### PR TITLE
Add SnoopPrecompile package

### PR DESCRIPTION
**Pull request by timholy**
This small package makes it easy to exhaustively precompile on
Julia 1.8 and higher. In brief, your package should look like

```julia
module MyPackage

using SnoopPrecompile

# method definitions etc

let
    @precompile_calls begin
        # execute the code you want precompiled
    end
end

end # module MyPackage
```

There are also facilities for "setup" code, see the documentation
for details.

### CODE REVIEW

**Comment by timholy** [comment]
CC @ChrisRackauckas, @SimonDanisch, @rikhuijzer, @LilithHafner, @IanButterworth. This gets both the inferrable and runtime-dispatched calls.

**Comment by rikhuijzer** [comment]
So basically this approach falls in between running a workload and calling `precompile` manually? Compared to the workload, it doesn't cause side-effects and compared to calling `precompile` it will also call runtime dispatches?

Why is `SnoopPrecompile` located inside the `SnoopCompile` repo and not in a separate one?

**Comment by LilithHafner** [comment]
>
```julia
let
    @precompile_calls begin
        # execute the code you want precompiled
    end
end
```

Wouldn't it be easier for the macro to introduce the let block?

File: SnoopPrecompile/Project.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name = "SnoopPrecompile"
+uuid = "66db9d55-30c0-4569-8b51-7e840670fc0c"
+authors = ["Tim Holy <tim.holy@gmail.com> and contributors"]
+version = "1.0.0"
+
+[compat]
+julia = "1"
```
[the remark this answers is not in the document]
Comment (timholy) [reply SnoopPrecompile/Project.toml:7]: I think of this as a "support" package (one that does not really exist in its own right but merely to make others better), and this allows it to support packages that happen to still support Julia 1.0. It won't do much useful on 1.0, but this way it doesn't force anyone to abandon 1.0 support.

**Comment by timholy** [comment]
> Wouldn't it be easier for the macro to introduce the let block?

Handling `:setup` code that way would have required
```julia
@precompile_calls :setup begin
    vars = ...
    @precompile calls begin
        somecall(vars...)
    end
end
```
since `vars` needs to be in scope for `somecall(vars...)`. I got the impression you didn't like the nested approach. But I can go that way if you prefer it.

**Comment by LilithHafner** [comment]
Using that approach, `@precompile_calls` without `:setup` could still let-fence itself.

Another approach uses only a single macro call, following the style of `BenchmarkTools.@benchmark`:
```julia
@precompile_calls begin
    somecalls(vars...)
end setup = (vars = ...)
```

**Comment by timholy** [comment]
> For the macro name @precompile_calls, I would suggest starting with a verb since most functions and macro's either start with a verb (e.g., @time, @benchmark and @debug) or state the output (e.g., @code_llvm and @lazy_str). Maybe @call_precompile?

It does :slightly_smiling_face: . I'm `precompile`ing the `calls` in that expression!

**Comment by timholy** [comment]
But of course happy to accept other names, given that it's obviously confusing!

**Comment by LilithHafner** [comment]
> It does 🙂 . I'm `precompile`ing the `calls` in that expression!

That is 100% obvious now that you say it, but I also didn't realize that precompile was a verb in that context.

File: docs/src/snoop_pc.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+struct MyType
+    x::Int
+end
+struct OtherType
+    str::String
+end
+
+let  # `let` prevents `list` from being a visible private name in the package
+    @precompile_calls :setup beginNaturally, y
```
Comment (jkrumbiegel) [inline docs/src/snoop_pc.md:24]: What is `beginNaturally, y` here?

Comment (timholy) [reply docs/src/snoop_pc.md:24]: Weird vscode paste! Should just be `begin`. Every once in a while vscode does weird things to my text, I must be sloppy about exiting multicursor or something?

**Comment by timholy** [comment]
If a name change doesn't seem in the offing, clearly it needs at least a docstring.

**Comment by timholy** [comment]
I think I've addressed most of the feedback. Some comments:

- I was worried that there could be a large amount of setup code, so the BenchmarkTools `setup = (...)` syntax began to seem a little less appealing. I went with the nested approach. Alternatively we could have `@precompile_setup` and `@precompile_calls`.
- Sadly, we seem to have a bug in our ability to disable the interpreter, so automatic `let`-fencing is not yet possible. Until then, you can use manual fencing at top level, but be sure to check results with `verbose[] = true`.

**Comment by timholy** [comment]
I'm beginning to lean towards `@precompile_setup` and `@precompile_all`, since people found `@precompile_calls` confusing. Thoughts?

**Comment by rikhuijzer** [comment]
+1 for `@precompile_all`

**Comment by jkrumbiegel** [comment]
How about `@precompile_all_calls`, it's longer but clear in that calls are the object not the verb, and otherwise you don't know from reading what's `all`.

**Comment by timholy** [comment]
I went with `@precompile_setup` and `@precompile_all_calls`. Barring any additional comments I will merge and register the package.

**Comment by timholy** [comment]
https://github.com/JuliaRegistries/General/pull/64746

**Pull request merged**